cairo/ethereum/cancun/keth/aggregator.cairo

```diff
@@ -0,0 +1,224 @@
+    alloc_locals;
+
+    // Number of body chunks executed.
+    local n_body_chunks: felt;
+    // Program hashes of the Keth segments.
+    local init_program_hash: felt;
+    local body_program_hash: felt;
+    local teardown_program_hash: felt;
+
+    // Pointers to the *serialized* actual outputs of each segment.
+    local serialized_init_output: felt*;
+    local serialized_body_outputs: felt**;
+    local serialized_teardown_output: felt*;
+
+    %{aggregator_inputs%}
+
+    with_attr error_message("AssertionError: Must have at least one body chunk") {
+        assert_not_equal(n_body_chunks, 0);
+    }
+
+    let init_output: KethInitOutput* = cast(serialized_init_output, KethInitOutput*);
+    let teardown_output: KethTeardownOutput* = cast(
+        serialized_teardown_output, KethTeardownOutput*
+    );
+
+    // --- Verify Commitments ---
+    // 1. Check init output links to the first body input
+    let first_body_output: KethBodyOutput* = cast(serialized_body_outputs[0], KethBodyOutput*);
+    assert init_output.body_commitment_0 = first_body_output.input_commitment_0;
+    assert init_output.body_commitment_1 = first_body_output.input_commitment_1;
+
+    // 2. Check body chunk links recursively/iteratively
+    // Only check internal links if there are more than 1 body chunk
+    let squared_n_body_chunks = n_body_chunks * n_body_chunks;
+    if (squared_n_body_chunks != n_body_chunks) {
+        check_body_chunk_links(
+            body_outputs_ptr_array=serialized_body_outputs,
+            current_chunk_index=0,
+            n_body_chunks=n_body_chunks,
+        );
+    }
+
+    // 3. Check last body output links to teardown input
+    let last_body_output: KethBodyOutput* = cast(
+        serialized_body_outputs[n_body_chunks - 1], KethBodyOutput*
+    );
+    assert last_body_output.post_exec_commitment_0 = teardown_output.body_args_commitment_check_0;
+    assert last_body_output.post_exec_commitment_1 = teardown_output.body_args_commitment_check_1;
+
+    // 4. Check init output links to teardown input
+    assert init_output.teardown_commitment_0 = teardown_output.init_args_commitment_check_0;
+    assert init_output.teardown_commitment_1 = teardown_output.init_args_commitment_check_1;
+
+    // --- Construct Output ---
+    // Write the output in the format expected by ApplicativeBootloader's memcpy check.
+    // This mimics the output of `run_bootloader` for plain tasks.
+
+    let output_ptr = write_segment_output(
+        output_ptr=output_ptr,
+        program_hash=init_program_hash,
+        serialized_output_data=serialized_init_output,
+        output_data_size=KethInitOutput.SIZE,
+    );
+
+    let output_ptr = write_body_segment_outputs(
+        output_ptr=output_ptr,
+        program_hash=body_program_hash,
+        serialized_body_outputs_array=serialized_body_outputs,
+        n_body_chunks=n_body_chunks,
+        current_chunk_index=0,
+    );
+
+    let output_ptr = write_segment_output(
```

Comment:
are these the steps they described as 'squash'? 
because if we write all outputs then there is no squashing?